File: python/geny_harness/tools/registry.py
"""Tool registry — registration, lookup, preset management."""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Set

from geny_harness.tools.base import Tool


class ToolRegistry:
    """Central registry for tools.

    Supports:
      - Registration by name
      - Preset-based filtering
      - API format export
    """

    def __init__(self) -> None:
        self._tools: Dict[str, Tool] = {}

    def register(self, tool: Tool) -> ToolRegistry:
        """Register a tool.  Chaining supported."""
        self._tools[tool.name] = tool
        return self

    def unregister(self, name: str) -> None:
        """Remove a tool by name."""
        self._tools.pop(name, None)

    def get(self, name: str) -> Optional[Tool]:
        """Get a tool by name."""
        return self._tools.get(name)

    def list_all(self) -> List[Tool]:
        """List all registered tools."""
        return list(self._tools.values())

    def list_names(self) -> List[str]:
        """List all registered tool names."""
        return list(self._tools.keys())

    def filter(
        self,
        include: Optional[Set[str]] = None,
        exclude: Optional[Set[str]] = None,
    ) -> List[Tool]:
        """Filter tools by name sets."""
        tools = self.list_all()
        if include is not None:
            tools = [t for t in tools if t.name in include]
        if exclude is not None:
            tools = [t for t in tools if t.name not in exclude]
        return tools

    def to_api_format(
        self,
        include: Optional[Set[str]] = None,
        exclude: Optional[Set[str]] = None,
    ) -> List[Dict[str, Any]]:
        """Export tools in Anthropic API format."""
        tools = self.filter(include=include, exclude=exclude)
        return [t.to_api_format() for t in tools]

    def __len__(self) -> int:
        return len(self._tools)

    def __contains__(self, name: str) -> bool:
        return name in self._tools

File: crates/geny-harness-core/src/tools/registry.rs
//! Tool registration and lookup.

use serde_json::Value;
use std::collections::HashSet;

use crate::tools::base::Tool;

/// Central registry for tools.
pub struct ToolRegistry {
    tools: Vec<Box<dyn Tool>>,
}

impl Default for ToolRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl ToolRegistry {
    pub fn new() -> Self {
        Self { tools: Vec::new() }
    }

    /// Register a tool.
    pub fn register(&mut self, tool: Box<dyn Tool>) -> &mut Self {
        self.tools.push(tool);
        self
    }

    /// Remove a tool by name.
    pub fn unregister(&mut self, name: &str) {
        self.tools.retain(|t| t.name() != name);
    }

    /// Lookup by name.
    pub fn get(&self, name: &str) -> Option<&dyn Tool> {
        self.tools
            .iter()
            .find(|t| t.name() == name)
            .map(|t| t.as_ref())
    }

    /// All registered tools.
    pub fn list_all(&self) -> Vec<&dyn Tool> {
        self.tools.iter().map(|t| t.as_ref()).collect()
    }

    /// All tool names.
    pub fn list_names(&self) -> Vec<&str> {
        self.tools.iter().map(|t| t.name()).collect()
    }

    /// Filter tools by include/exclude name sets.
    pub fn filter(
        &self,
        include: Option<&HashSet<String>>,
        exclude: Option<&HashSet<String>>,
    ) -> Vec<&dyn Tool> {
        self.tools
            .iter()
            .filter(|t| {
                if let Some(inc) = include {
                    if !inc.contains(t.name()) {
                        return false;
                    }
                }
                if let Some(exc) = exclude {
                    if exc.contains(t.name()) {
                        return false;
                    }
                }
                true
            })
            .map(|t| t.as_ref())
            .collect()
    }

    /// Export filtered tools in API format.
    pub fn to_api_format(
        &self,
        include: Option<&HashSet<String>>,
        exclude: Option<&HashSet<String>>,
    ) -> Vec<Value> {
        self.filter(include, exclude)
            .iter()
            .map(|t| t.to_api_format())
            .collect()
    }

    /// Tool count.
    pub fn len(&self) -> usize {
        self.tools.len()
    }

    /// Check if empty.
    pub fn is_empty(&self) -> bool {
        self.tools.is_empty()
    }

    /// Check if a tool name is registered.
    pub fn contains(&self, name: &str) -> bool {
        self.tools.iter().any(|t| t.name() == name)
    }
}

impl std::fmt::Debug for ToolRegistry {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ToolRegistry")
            .field("count", &self.tools.len())
            .field("names", &self.list_names())
            .finish()
    }
}
